File: DailyChallenges/Scala/LeetCode_133.scala
object Solution {
    def cloneGraph(graph: Node): Node = {
        val m = collection.mutable.HashMap.empty[Int, Node]
        def clone(node: Node): Node = {
            if (node == null) null else if (m.contains(node.value)) m(node.value) else {
                val nn = new Node(node.value)
                m(node.value) = nn
                for (n <- node.neighbors.reverse)
                    nn.neighbors = clone(n) :: nn.neighbors
                nn
            }
        }
        clone(graph)
    }
}

File: DailyChallenges/CPP/LeetCode_133.cpp
class Solution {
public:
    unordered_map<int, Node *> m;
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return node;
        if (m[node->val] != nullptr) return m[node->val];
        Node * nn = new Node(node->val);
        m[nn->val] = nn;
        for (auto child: node->neighbors)
            nn->neighbors.push_back(cloneGraph(child));
        return nn; 
    }
};
